// @eclipse-glsp/client/lib/features/accessibility/search/search-tool.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var SearchAutocompletePaletteTool_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchAutocompletePaletteKeyListener = exports.SearchAutocompletePaletteTool = void 0;
const inversify_1 = require("inversify");
const keyboard_1 = require("sprotty/lib/utils/keyboard");
const sprotty_1 = require("@eclipse-glsp/sprotty");
const action_dispatcher_1 = require("../../../base/action-dispatcher");
const accessible_key_shortcut_1 = require("../key-shortcut/accessible-key-shortcut");
const search_palette_1 = require("./search-palette");
let SearchAutocompletePaletteTool = SearchAutocompletePaletteTool_1 = class SearchAutocompletePaletteTool {
    constructor() {
        this.keyListener = new SearchAutocompletePaletteKeyListener(this);
    }
    get id() {
        return SearchAutocompletePaletteTool_1.ID;
    }
    enable() {
        this.keyTool.register(this.keyListener);
        this.keyListener.registerShortcutKey();
    }
    disable() {
        this.keyTool.deregister(this.keyListener);
    }
};
exports.SearchAutocompletePaletteTool = SearchAutocompletePaletteTool;
SearchAutocompletePaletteTool.ID = 'glsp.search-autocomplete-palette-tool';
__decorate([
    (0, inversify_1.inject)(sprotty_1.TYPES.IActionDispatcher),
    __metadata("design:type", action_dispatcher_1.GLSPActionDispatcher)
], SearchAutocompletePaletteTool.prototype, "actionDispatcher", void 0);
__decorate([
    (0, inversify_1.inject)(sprotty_1.KeyTool),
    __metadata("design:type", sprotty_1.KeyTool)
], SearchAutocompletePaletteTool.prototype, "keyTool", void 0);
exports.SearchAutocompletePaletteTool = SearchAutocompletePaletteTool = SearchAutocompletePaletteTool_1 = __decorate([
    (0, inversify_1.injectable)()
], SearchAutocompletePaletteTool);
class SearchAutocompletePaletteKeyListener extends sprotty_1.KeyListener {
    constructor(tool) {
        super();
        this.tool = tool;
        this.token = search_palette_1.SearchAutocompletePalette.name;
    }
    registerShortcutKey() {
        this.tool.actionDispatcher.dispatchOnceModelInitialized(accessible_key_shortcut_1.SetAccessibleKeyShortcutAction.create({
            token: this.token,
            keys: [{ shortcuts: ['CTRL', 'F'], description: 'Activate search for elements', group: 'Search', position: 0 }]
        }));
    }
    keyDown(element, event) {
        if (this.matchesSearchActivateKeystroke(event)) {
            return [
                sprotty_1.SetUIExtensionVisibilityAction.create({
                    extensionId: search_palette_1.SearchAutocompletePalette.ID,
                    visible: true
                })
            ];
        }
        return [];
    }
    matchesSearchActivateKeystroke(event) {
        return (0, keyboard_1.matchesKeystroke)(event, 'KeyF', 'ctrl');
    }
}
exports.SearchAutocompletePaletteKeyListener = SearchAutocompletePaletteKeyListener;
