// @eclipse-glsp/client/lib/features/validation/marker-navigator.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2020-2023 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const sprotty_1 = require("@eclipse-glsp/sprotty");
const chai_1 = require("chai");
const inversify_1 = require("inversify");
require("mocha");
require("reflect-metadata");
const default_module_1 = require("../../base/default.module");
const model_1 = require("../../model");
const decoration_module_1 = require("../decoration/decoration-module");
const issue_marker_1 = require("./issue-marker");
const marker_navigator_1 = require("./marker-navigator");
const validation_modules_1 = require("./validation-modules");
describe('MarkerNavigator', () => {
    const container = new inversify_1.Container();
    container.load(default_module_1.defaultModule, decoration_module_1.decorationModule, validation_modules_1.markerNavigatorModule);
    const markerNavigator = container.get(marker_navigator_1.MarkerNavigator);
    function createRoot(...nodeIds) {
        const root = new model_1.GGraph();
        root.features = new Set(model_1.GGraph.DEFAULT_FEATURES);
        root.id = 'root';
        root.type = 'graph';
        nodeIds.forEach(id => {
            const node = new sprotty_1.GNode();
            node.id = id;
            node.type = 'node';
            node.features = new Set(sprotty_1.GNode.DEFAULT_FEATURES);
            root.add(node);
        });
        return root;
    }
    const rootWithoutAnyMarkers = createRoot('1');
    const rootWithMarkers = createRoot('bottom-right', 'top-right', 'top-left', 'bottom-left');
    const elementTopLeft = rootWithMarkers.children[2];
    elementTopLeft.bounds = { width: 10, height: 10, x: 100, y: 100 };
    const elementTopRight = rootWithMarkers.children[1];
    elementTopRight.bounds = { width: 10, height: 10, x: 200, y: 100 };
    const elementBottomLeft = rootWithMarkers.children[3];
    elementBottomLeft.bounds = { width: 10, height: 10, x: 100, y: 200 };
    const elementBottomRight = rootWithMarkers.children[0];
    elementBottomRight.bounds = { width: 10, height: 10, x: 200, y: 200 };
    beforeEach('clear issue marker', () => {
        [elementTopLeft, elementTopRight, elementBottomLeft, elementBottomRight].forEach(clearMarker);
    });
    it('next(undefined) without any markers returns undefined', () => {
        (0, chai_1.expect)(markerNavigator.next(rootWithoutAnyMarkers)).to.be.undefined;
    });
    it('previous(undefined) without any markers returns undefined', () => {
        (0, chai_1.expect)(markerNavigator.previous(rootWithoutAnyMarkers)).to.be.undefined;
    });
    it('next(undefined) with one marker returns the one marker', () => {
        const marker = setIssues(elementTopLeft, [{ message: 'msg', severity: 'error' }]);
        const next = markerNavigator.next(rootWithMarkers);
        (0, chai_1.expect)(next).to.eql(marker);
    });
    it('next(firstMarker) with only one marker returns again the first marker', () => {
        const marker = setIssues(elementTopLeft, [{ message: 'msg', severity: 'error' }]);
        const next = markerNavigator.next(rootWithMarkers, marker);
        (0, chai_1.expect)(next).to.eql(marker);
        // and again and again
        const nextNext = markerNavigator.next(rootWithMarkers, next);
        (0, chai_1.expect)(nextNext).to.eql(marker);
    });
    it('previous(firstMarker) with only one marker returns again the first marker', () => {
        const marker = setIssues(elementTopLeft, [{ message: 'msg', severity: 'error' }]);
        const previous = markerNavigator.previous(rootWithMarkers, marker);
        (0, chai_1.expect)(previous).to.eql(marker);
        // and again and again
        const previousPrevious = markerNavigator.previous(rootWithMarkers, previous);
        (0, chai_1.expect)(previousPrevious).to.eql(marker);
    });
    it('next(firstMarker) with two marker returns second marker then again first marker', () => {
        const marker1 = setIssues(elementTopLeft, [{ message: 'msg', severity: 'error' }]);
        const marker2 = setIssues(elementTopRight, [{ message: 'msg', severity: 'error' }]);
        const next = markerNavigator.next(rootWithMarkers, marker1);
        (0, chai_1.expect)(next).to.eql(marker2);
        // and again and again
        const nextNext = markerNavigator.next(rootWithMarkers, next);
        (0, chai_1.expect)(nextNext).to.eql(marker1);
    });
    it('previous(firstMarker) with two marker returns second marker then again first marker', () => {
        const marker1 = setIssues(elementTopLeft, [{ message: 'msg', severity: 'error' }]);
        const marker2 = setIssues(elementTopRight, [{ message: 'msg', severity: 'error' }]);
        const next = markerNavigator.previous(rootWithMarkers, marker1);
        (0, chai_1.expect)(next).to.eql(marker2);
        // and again and again
        const nextNext = markerNavigator.previous(rootWithMarkers, next);
        (0, chai_1.expect)(nextNext).to.eql(marker1);
    });
    it('returns markers in the order left-to-right, top-to-bottom with next()', () => {
        const marker1 = setIssues(elementTopLeft, [{ message: 'msg', severity: 'error' }]);
        const marker2 = setIssues(elementTopRight, [{ message: 'msg', severity: 'error' }]);
        const marker3 = setIssues(elementBottomLeft, [{ message: 'msg', severity: 'error' }]);
        const marker4 = setIssues(elementBottomRight, [{ message: 'msg', severity: 'error' }]);
        const found1 = markerNavigator.next(rootWithMarkers);
        const found2 = markerNavigator.next(rootWithMarkers, found1);
        const found3 = markerNavigator.next(rootWithMarkers, found2);
        const found4 = markerNavigator.next(rootWithMarkers, found3);
        const found5 = markerNavigator.next(rootWithMarkers, found4);
        (0, chai_1.expect)(found1).to.eql(marker1);
        (0, chai_1.expect)(found2).to.eql(marker2);
        (0, chai_1.expect)(found3).to.eql(marker3);
        (0, chai_1.expect)(found4).to.eql(marker4);
        (0, chai_1.expect)(found5).to.eql(marker1);
    });
    it('returns markers in the order left-to-right, top-to-bottom with previous()', () => {
        const marker1 = setIssues(elementTopLeft, [{ message: 'msg', severity: 'error' }]);
        const marker2 = setIssues(elementTopRight, [{ message: 'msg', severity: 'error' }]);
        const marker3 = setIssues(elementBottomLeft, [{ message: 'msg', severity: 'error' }]);
        const marker4 = setIssues(elementBottomRight, [{ message: 'msg', severity: 'error' }]);
        const found1 = markerNavigator.previous(rootWithMarkers);
        const found2 = markerNavigator.previous(rootWithMarkers, found1);
        const found3 = markerNavigator.previous(rootWithMarkers, found2);
        const found4 = markerNavigator.previous(rootWithMarkers, found3);
        const found5 = markerNavigator.previous(rootWithMarkers, found4);
        (0, chai_1.expect)(found1).to.eql(marker1);
        (0, chai_1.expect)(found2).to.eql(marker4);
        (0, chai_1.expect)(found3).to.eql(marker3);
        (0, chai_1.expect)(found4).to.eql(marker2);
        (0, chai_1.expect)(found5).to.eql(marker1);
    });
});
function clearMarker(elem) {
    elem.children.filter(isMarker).forEach(marker => elem.remove(marker));
}
function setIssues(elem, issues) {
    const marker = getOrCreateMarker(elem);
    marker.issues = issues;
    return marker;
}
function getOrCreateMarker(elem) {
    const marker = findMarker(elem);
    if (marker instanceof issue_marker_1.GIssueMarker) {
        return marker;
    }
    return createMarker(elem);
}
function findMarker(elem) {
    return elem.children.find(isMarker);
}
function isMarker(element) {
    return element instanceof issue_marker_1.GIssueMarker;
}
function createMarker(elem) {
    const newMarker = new issue_marker_1.GIssueMarker();
    newMarker.type = 'marker';
    newMarker.id = `${elem.id}_marker`;
    elem.add(newMarker);
    return newMarker;
}
