// @eclipse-glsp/protocol/lib/action-protocol/client-notification.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022-2023 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const client_notification_1 = require("./client-notification");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Client notification actions', () => {
    describe('StatusAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const statusAction = { kind: client_notification_1.StatusAction.KIND, message: 'Some', severity: 'INFO' };
                (0, chai_1.expect)(client_notification_1.StatusAction.is(statusAction)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(client_notification_1.StatusAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(client_notification_1.StatusAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const message = 'someMessage';
                const expected = { kind: client_notification_1.StatusAction.KIND, message, severity: 'INFO' };
                (0, chai_1.expect)(client_notification_1.StatusAction.create(message)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: client_notification_1.StatusAction.KIND,
                    message: 'someMessage',
                    severity: 'ERROR',
                    timeout: 5
                };
                const { message, severity, timeout } = expected;
                (0, chai_1.expect)(client_notification_1.StatusAction.create(message, { severity, timeout })).to.deep.equals(expected);
            });
        });
    });
    describe('MessageAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const messageAction = { kind: client_notification_1.MessageAction.KIND, message: '', severity: 'INFO' };
                (0, chai_1.expect)(client_notification_1.MessageAction.is(messageAction)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(client_notification_1.MessageAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(client_notification_1.MessageAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const message = 'someMessage';
                const expected = { kind: client_notification_1.MessageAction.KIND, message, severity: 'INFO' };
                (0, chai_1.expect)(client_notification_1.MessageAction.create(message)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: client_notification_1.MessageAction.KIND,
                    message: 'someMessage',
                    details: 'details',
                    severity: 'ERROR'
                };
                const { message, severity, details } = expected;
                (0, chai_1.expect)(client_notification_1.MessageAction.create(message, { severity, details })).to.deep.equals(expected);
            });
        });
    });
    describe('StartProgressAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const messageAction = { kind: 'startProgress', progressId: '1', title: 'Progress title' };
                (0, chai_1.expect)(client_notification_1.StartProgressAction.is(messageAction)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(client_notification_1.StartProgressAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(client_notification_1.StartProgressAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const progressId = '1';
                const title = 'Progress title';
                const expected = { kind: client_notification_1.StartProgressAction.KIND, progressId, title };
                (0, chai_1.expect)(client_notification_1.StartProgressAction.create({ progressId, title })).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: client_notification_1.StartProgressAction.KIND,
                    progressId: '1',
                    title: 'Progress title',
                    message: 'Some message',
                    percentage: 10
                };
                const { progressId, title, message, percentage } = expected;
                (0, chai_1.expect)(client_notification_1.StartProgressAction.create({ progressId, title, message, percentage })).to.deep.equals(expected);
            });
        });
    });
    describe('UpdateProgressAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const messageAction = { kind: 'updateProgress', progressId: '1' };
                (0, chai_1.expect)(client_notification_1.UpdateProgressAction.is(messageAction)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(client_notification_1.UpdateProgressAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(client_notification_1.UpdateProgressAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const progressId = '1';
                const expected = { kind: client_notification_1.UpdateProgressAction.KIND, progressId };
                (0, chai_1.expect)(client_notification_1.UpdateProgressAction.create(progressId)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: client_notification_1.UpdateProgressAction.KIND,
                    progressId: '1',
                    message: 'Some message',
                    percentage: 10
                };
                const { progressId, message, percentage } = expected;
                (0, chai_1.expect)(client_notification_1.UpdateProgressAction.create(progressId, { message, percentage })).to.deep.equals(expected);
            });
        });
    });
    describe('EndProgressAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const messageAction = { kind: 'endProgress', progressId: '1' };
                (0, chai_1.expect)(client_notification_1.EndProgressAction.is(messageAction)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(client_notification_1.EndProgressAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(client_notification_1.EndProgressAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const progressId = '1';
                const expected = { kind: client_notification_1.EndProgressAction.KIND, progressId, message: undefined };
                (0, chai_1.expect)(client_notification_1.EndProgressAction.create(progressId)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: client_notification_1.EndProgressAction.KIND,
                    progressId: '1',
                    message: 'Some message'
                };
                const { progressId, message } = expected;
                (0, chai_1.expect)(client_notification_1.EndProgressAction.create(progressId, message)).to.deep.equals(expected);
            });
        });
    });
});
