// @eclipse-glsp/client/lib/features/tools/node-creation/node-creation-tool.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var NodeCreationTool_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeCreationToolMouseListener = exports.NodeCreationTool = void 0;
/********************************************************************************
 * Copyright (c) 2020-2023 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const inversify_1 = require("inversify");
const sprotty_1 = require("@eclipse-glsp/sprotty");
const drag_aware_mouse_listener_1 = require("../../../base/drag-aware-mouse-listener");
const css_feedback_1 = require("../../../base/feedback/css-feedback");
const tool_1 = require("../../../base/tool-manager/tool");
const viewpoint_util_1 = require("../../../utils/viewpoint-util");
const model_1 = require("../../hints/model");
const base_tools_1 = require("../base-tools");
let NodeCreationTool = NodeCreationTool_1 = class NodeCreationTool extends base_tools_1.BaseCreationTool {
    constructor() {
        super(...arguments);
        this.isTriggerAction = sprotty_1.TriggerNodeCreationAction.is;
    }
    get id() {
        return NodeCreationTool_1.ID;
    }
    doEnable() {
        this.toDisposeOnDisable.push(this.mouseTool.registerListener(new NodeCreationToolMouseListener(this.triggerAction, this)), this.registerFeedback([(0, css_feedback_1.cursorFeedbackAction)(css_feedback_1.CursorCSS.NODE_CREATION)], this, [(0, css_feedback_1.cursorFeedbackAction)()]));
    }
};
exports.NodeCreationTool = NodeCreationTool;
NodeCreationTool.ID = 'tool_create_node';
__decorate([
    (0, inversify_1.inject)(sprotty_1.TYPES.ISnapper),
    (0, inversify_1.optional)(),
    __metadata("design:type", Object)
], NodeCreationTool.prototype, "snapper", void 0);
exports.NodeCreationTool = NodeCreationTool = NodeCreationTool_1 = __decorate([
    (0, inversify_1.injectable)()
], NodeCreationTool);
let NodeCreationToolMouseListener = class NodeCreationToolMouseListener extends drag_aware_mouse_listener_1.DragAwareMouseListener {
    constructor(triggerAction, tool) {
        super();
        this.triggerAction = triggerAction;
        this.tool = tool;
    }
    creationAllowed(elementTypeId) {
        return this.container && this.container.isContainableElement(elementTypeId);
    }
    get elementTypeId() {
        return this.triggerAction.elementTypeId;
    }
    nonDraggingMouseUp(target, event) {
        const result = [];
        if (this.container === undefined) {
            this.mouseOver(target, event);
        }
        if (this.creationAllowed(this.elementTypeId)) {
            const containerId = this.container ? this.container.id : undefined;
            let location = (0, viewpoint_util_1.getAbsolutePosition)(target, event);
            if (this.tool.snapper) {
                // Create a 0-bounds proxy element for snapping
                const elementProxy = new sprotty_1.GNode();
                elementProxy.size = { width: 0, height: 0 };
                location = this.tool.snapper.snap(location, elementProxy);
            }
            result.push(sprotty_1.CreateNodeOperation.create(this.elementTypeId, { location, containerId, args: this.triggerAction.args }));
            if (!(0, sprotty_1.isCtrlOrCmd)(event)) {
                result.push(tool_1.EnableDefaultToolsAction.create());
            }
        }
        return result;
    }
    mouseOver(target, event) {
        const currentContainer = (0, sprotty_1.findParentByFeature)(target, model_1.isContainable);
        if (!this.container || currentContainer !== this.container) {
            this.container = currentContainer;
            const feedback = this.creationAllowed(this.elementTypeId)
                ? (0, css_feedback_1.cursorFeedbackAction)(css_feedback_1.CursorCSS.NODE_CREATION)
                : (0, css_feedback_1.cursorFeedbackAction)(css_feedback_1.CursorCSS.OPERATION_NOT_ALLOWED);
            this.tool.registerFeedback([feedback]);
        }
        return [];
    }
};
exports.NodeCreationToolMouseListener = NodeCreationToolMouseListener;
exports.NodeCreationToolMouseListener = NodeCreationToolMouseListener = __decorate([
    (0, inversify_1.injectable)(),
    __metadata("design:paramtypes", [Object, NodeCreationTool])
], NodeCreationToolMouseListener);
