// @eclipse-glsp/client/lib/features/select/select-mouse-listener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RankedSelectMouseListener = void 0;
/********************************************************************************
 * Copyright (c) 2019-2023 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const sprotty_1 = require("@eclipse-glsp/sprotty");
const ranked_1 = require("../../base/ranked");
/**
 * Ranked select mouse listener that is executed before default mouse listeners when using the RankedMouseTool.
 * This ensures that default mouse listeners are working on a model that has selection changes already applied.
 */
class RankedSelectMouseListener extends sprotty_1.SelectMouseListener {
    constructor() {
        super(...arguments);
        this.rank = ranked_1.Ranked.DEFAULT_RANK - 100; /* we want to be executed before all default mouse listeners */
    }
    handleSelectTarget(selectableTarget, deselectedElements, event) {
        const result = [];
        result.push(sprotty_1.SelectAction.create({
            selectedElementsIDs: [selectableTarget.id],
            deselectedElementsIDs: deselectedElements.map(e => e.id)
        }));
        result.push(sprotty_1.BringToFrontAction.create([selectableTarget.id]));
        return result;
    }
    handleDeselectTarget(selectableTarget, event) {
        const result = [];
        result.push(sprotty_1.SelectAction.create({ selectedElementsIDs: [], deselectedElementsIDs: [selectableTarget.id] }));
        return result;
    }
    handleDeselectAll(deselectedElements, event) {
        const result = [];
        result.push(sprotty_1.SelectAction.create({ selectedElementsIDs: [], deselectedElementsIDs: deselectedElements.map(e => e.id) }));
        return result;
    }
}
exports.RankedSelectMouseListener = RankedSelectMouseListener;
