// @eclipse-glsp/client/lib/features/accessibility/resize-key-tool/resize-key-handler.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var ResizeElementHandler_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResizeElementHandler = exports.ResizeElementAction = exports.ResizeType = void 0;
const inversify_1 = require("inversify");
const sprotty_1 = require("@eclipse-glsp/sprotty");
const editor_context_service_1 = require("../../../base/editor-context-service");
const layout_utils_1 = require("../../../utils/layout-utils");
const gmodel_util_1 = require("../../../utils/gmodel-util");
const snap_1 = require("../../change-bounds/snap");
var ResizeType;
(function (ResizeType) {
    ResizeType[ResizeType["Increase"] = 0] = "Increase";
    ResizeType[ResizeType["Decrease"] = 1] = "Decrease";
    ResizeType[ResizeType["MinSize"] = 2] = "MinSize";
})(ResizeType || (exports.ResizeType = ResizeType = {}));
var ResizeElementAction;
(function (ResizeElementAction) {
    ResizeElementAction.KIND = 'resizeElementAction';
    function is(object) {
        return sprotty_1.Action.hasKind(object, ResizeElementAction.KIND);
    }
    ResizeElementAction.is = is;
    function create(elementIds, resizeType) {
        return { kind: ResizeElementAction.KIND, elementIds, resizeType };
    }
    ResizeElementAction.create = create;
})(ResizeElementAction || (exports.ResizeElementAction = ResizeElementAction = {}));
/* The ResizeElementHandler class is an implementation of the IActionHandler interface that handles
resizing of elements. */
let ResizeElementHandler = ResizeElementHandler_1 = class ResizeElementHandler {
    constructor(snapper) {
        this.snapper = snapper;
        this.grid = { x: ResizeElementHandler_1.defaultResizeX, y: ResizeElementHandler_1.defaultResizeY };
        this.isEditMode = false;
        if (snapper instanceof snap_1.GridSnapper) {
            this.grid = snapper.grid;
        }
    }
    handle(action) {
        if (ResizeElementAction.is(action)) {
            this.handleResizeElement(action);
        }
    }
    handleResizeElement(action) {
        const elements = (0, gmodel_util_1.getElements)(this.editorContextService.modelRoot.index, action.elementIds, gmodel_util_1.isSelectableAndBoundsAware);
        this.dispatcher.dispatchAll(this.resize(elements, action));
    }
    resize(elements, action) {
        const actions = [];
        elements.forEach(element => {
            const { x, y, width: oldWidth, height: oldHeight } = element.bounds;
            let width = 0;
            let height = 0;
            if (action.resizeType === ResizeType.Decrease) {
                width = oldWidth - this.grid.x;
                height = oldHeight - this.grid.y;
            }
            else if (action.resizeType === ResizeType.Increase) {
                width = oldWidth + this.grid.x;
                height = oldHeight + this.grid.y;
            }
            else if (action.resizeType === ResizeType.MinSize) {
                width = (0, layout_utils_1.minWidth)(element);
                height = (0, layout_utils_1.minHeight)(element);
            }
            if (this.isValidBoundChange(element, { x, y }, { width, height })) {
                const resizeElement = { id: element.id, bounds: { x, y, width, height } };
                actions.push(sprotty_1.ChangeBoundsOperation.create([(0, gmodel_util_1.toElementAndBounds)(resizeElement)]));
            }
        });
        return actions;
    }
    isValidBoundChange(element, newPosition, newSize) {
        return (0, layout_utils_1.isValidSize)(element, newSize) && (0, layout_utils_1.isValidMove)(element, newPosition);
    }
};
exports.ResizeElementHandler = ResizeElementHandler;
// Default x resize used if GridSnapper is not provided
ResizeElementHandler.defaultResizeX = 20;
// Default y resize used if GridSnapper is not provided
ResizeElementHandler.defaultResizeY = 20;
__decorate([
    (0, inversify_1.inject)(editor_context_service_1.EditorContextService),
    __metadata("design:type", editor_context_service_1.EditorContextService)
], ResizeElementHandler.prototype, "editorContextService", void 0);
__decorate([
    (0, inversify_1.inject)(sprotty_1.TYPES.IActionDispatcher),
    __metadata("design:type", Object)
], ResizeElementHandler.prototype, "dispatcher", void 0);
exports.ResizeElementHandler = ResizeElementHandler = ResizeElementHandler_1 = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(sprotty_1.TYPES.ISnapper)),
    __param(0, (0, inversify_1.optional)()),
    __metadata("design:paramtypes", [Object])
], ResizeElementHandler);
