// @eclipse-glsp/theia-integration/lib/browser/glsp-theia-container-module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLSPTheiaFrontendModule = void 0;
/********************************************************************************
 * Copyright (C) 2021-2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const client_1 = require("@eclipse-glsp/client");
const inversify_1 = require("@theia/core/shared/inversify");
const _1 = require(".");
const copy_paste_context_menu_contribution_1 = require("./copy-paste-context-menu-contribution");
const diagram_widget_factory_1 = require("./diagram/diagram-widget-factory");
const glsp_diagram_manager_1 = require("./diagram/glsp-diagram-manager");
const glsp_layout_commands_1 = require("./diagram/glsp-layout-commands");
const glsp_client_contribution_1 = require("./glsp-client-contribution");
const theia_navigate_to_marker_contribution_1 = require("./theia-navigate-to-marker-contribution");
/**
 *  The `GLSPTheiaFrontendModule` provides the necessary required module configuration
 *  to implement a theia GLSP diagram language integration.
 */
class GLSPTheiaFrontendModule extends inversify_1.ContainerModule {
    constructor() {
        super((bind, unbind, isBound, rebind) => this.initialize({ bind, unbind, isBound, rebind }));
        this.enableLayoutCommands = true;
        this.enableMarkerNavigationCommands = true;
        this.enableCopyPaste = false;
    }
    initialize(context) {
        this.bindGLSPClientContribution(context);
        this.bindDiagramConfiguration(context);
        this.bindDiagramWidgetFactory(context);
        this.configureDiagramManager(context);
        // Optional default configuration
        this.configureDiagramLayoutCommands(context);
        this.configureCopyPasteContextMenu(context);
        this.configureMarkerNavigationCommands(context);
        this.configure(context);
    }
    /**
     * Defines the binding for the glsp client contribution of the diagram integration. A {@link ConfigurableGLSPClientContribution}
     * with the same id as the diagramLanguage is bound. Can be overwritten in subclasses to provide a custom binding using
     * the {@link GLSPClientContribution} service identifier.
     *
     * For example:
     * ```typescript
     * context.bind(GLSPClientContribution).toSelf(MyCustomGLSPClientContribution);
     * ```
     * Note that glsp client contribution bindings are consumed via multi-injection. This means binding the {@link GLSPClientContribution}
     * service identifier in singleton scope has no effect.
     *
     * @param context the container context
     */
    bindGLSPClientContribution(context) {
        context.bind(glsp_client_contribution_1.GLSPClientContribution).toDynamicValue(dynamicContext => {
            const contribution = dynamicContext.container.resolve(ConfigurableGLSPClientContribution);
            contribution.id = this.diagramLanguage.contributionId;
            return contribution;
        });
    }
    /**
     * Defines the binding for the {@link DiagramWidgetFactory} of the diagram integration. Per default a
     * factory that constructs a {@link GLSPDiagramWidget} instance is bound.
     * Can be overwritten in subclasses to provide a custom binding using
     * the {@link DiagramWidgetFactory} service identifier.
     *
     * For example:
     * ```typescript
     * context.bind(DiagramWidgetFactory).to(MyCustomDiagramWidgetFactory);
     * ```
     * Note that glsp client contribution bindings are consumed via multi-injection. This means binding the {@link GLSPClientContribution}
     * service identifier in singleton scope has no effect.
     *
     * @param context the container context
     */
    bindDiagramWidgetFactory(context) {
        var _a;
        (_a = (0, client_1.lazyBind)(context, _1.GLSPDiagramWidget)) === null || _a === void 0 ? void 0 : _a.toSelf();
        context
            .bind(diagram_widget_factory_1.DiagramWidgetFactory)
            .toDynamicValue(ctx => (0, diagram_widget_factory_1.createDiagramWidgetFactory)(ctx, this.diagramLanguage.diagramType))
            .inSingletonScope();
    }
    /**
     * Configures the bindings for the diagram manager of the diagram integration. A {@link ConfigurableGLSPDiagramManager}
     * is bound to a generated service identifier in singleton scope and then additional bindings for this service
     * identifier are registered. Can be overwritten in subclasses to provide a custom binding.
     *
     *  For example:
     * ```typescript
     * context.bind(MyDiagramManager).to(MyDiagramManager).inSingletonScope();
     * registerDiagramManager(context.bind, MyDiagramManager);
     * ```
     * @param context the container context
     */
    configureDiagramManager(context) {
        const diagramManagerServiceId = Symbol(`DiagramManager_${this.diagramLanguage.diagramType}`);
        context
            .bind(diagramManagerServiceId)
            .toDynamicValue(dynamicContext => {
            const manager = dynamicContext.container.resolve(ConfigurableGLSPDiagramManager);
            manager.doConfigure(this.diagramLanguage);
            return manager;
        })
            .inSingletonScope();
        (0, glsp_diagram_manager_1.registerDiagramManager)(context.bind, diagramManagerServiceId, false);
    }
    /**
     * Can be overwritten by subclasses to provide additional configuration (e.g. extra bindings)
     * for this module.
     *
     * For example:
     * ```typescript
     * context.bind(MyCustomClass).toSelf().inSingletonScope()
     * ```
     * @param context the container context
     */
    configure(context) {
        // Empty per default
    }
    /**
     * Optional configuration to enable the default diagram layout commands for the diagram integration.
     * Can be enabled/disabled using the {@link GLSPTheiaFrontendModule.enableLayoutCommands}  property flag.
     *
     * @param context the diagram context
     */
    configureDiagramLayoutCommands(context) {
        if (this.enableLayoutCommands) {
            (0, glsp_layout_commands_1.registerDiagramLayoutCommands)(context.bind);
        }
    }
    /**
     * Optional configuration for copy & paste functionality of the diagram integration.
     * Can be enabled/disabled using the {@link GLSPTheiaFrontendModule.enableCopyPaste} property flag.
     * Note that the glsp server also needs to support copy & paste for this diagram configuration.
     *
     * @param context the diagram context
     */
    configureCopyPasteContextMenu(context) {
        if (this.enableCopyPaste) {
            (0, copy_paste_context_menu_contribution_1.registerCopyPasteContextMenu)(context.bind);
        }
    }
    /**
     * Optional configuration to enable marker navigation for the diagram integration.
     * Can be enabled/disabled using the {@link GLSPTheiaFrontendModule.enableMarkerNavigationCommands}  property flag.
     * Note that the glsp server also needs to support copy & paste for this diagram configuration.
     *
     * @param context the diagram context
     */
    configureMarkerNavigationCommands(context) {
        if (this.enableMarkerNavigationCommands) {
            (0, theia_navigate_to_marker_contribution_1.registerMarkerNavigationCommands)(context.bind);
        }
    }
}
exports.GLSPTheiaFrontendModule = GLSPTheiaFrontendModule;
/**
 * Internal class that is used in {@link GLSPTheiaFrontendModule.configureDiagramManager} to
 * bind a default implementation for `DiagramManager`. A custom `DiagramManager` should
 * never extend this class. Use {@link GLSPDiagramManager} instead.
 */
let ConfigurableGLSPDiagramManager = class ConfigurableGLSPDiagramManager extends glsp_diagram_manager_1.GLSPDiagramManager {
    constructor() {
        super(...arguments);
        this._fileExtensions = [];
        this._iconClass = (0, client_1.codiconCSSString)('type-hierarchy-sub');
    }
    doConfigure(diagramLanguage) {
        this._fileExtensions = diagramLanguage.fileExtensions;
        this._diagramType = diagramLanguage.diagramType;
        this._label = diagramLanguage.label;
        this._iconClass = diagramLanguage.iconClass || this._iconClass;
        this._contributionId = diagramLanguage.contributionId;
    }
    get fileExtensions() {
        return this._fileExtensions;
    }
    get diagramType() {
        if (!this._diagramType) {
            throw new Error('No diagramType has been set for this ConfigurableGLSPDiagramManager');
        }
        return this._diagramType;
    }
    get label() {
        return this._label;
    }
    get contributionId() {
        return this._contributionId;
    }
    get iconClass() {
        return this._iconClass;
    }
};
ConfigurableGLSPDiagramManager = __decorate([
    (0, inversify_1.injectable)()
], ConfigurableGLSPDiagramManager);
/**
 * Internal class that is used in {@link GLSPTheiaFrontendModule.bindGLSPClientContribution} to
 * bind a default implementation for {@link GLSPClientContribution}. A custom {@link GLSPClientContribution}  should
 * never extend this class. Use {@link BaseGLSPClientContribution} instead.
 */
let ConfigurableGLSPClientContribution = class ConfigurableGLSPClientContribution extends glsp_client_contribution_1.BaseGLSPClientContribution {
    get id() {
        if (!this._id) {
            throw new Error('No id has been set for this DefaultTheiaGLSPConnector');
        }
        return this._id;
    }
    set id(value) {
        if (this._id) {
            throw new Error(`A value for id has already been set. Cannot set new value'${value}'`);
        }
        this._id = value;
    }
};
ConfigurableGLSPClientContribution = __decorate([
    (0, inversify_1.injectable)()
], ConfigurableGLSPClientContribution);
