// @eclipse-glsp/client/lib/features/tools/marquee-selection/marquee-mouse-tool.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var MarqueeMouseTool_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShiftKeyListener = exports.MarqueeMouseListener = exports.MarqueeMouseTool = void 0;
/********************************************************************************
 * Copyright (c) 2021-2023 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const sprotty_1 = require("@eclipse-glsp/sprotty");
const inversify_1 = require("inversify");
const drag_aware_mouse_listener_1 = require("../../../base/drag-aware-mouse-listener");
const css_feedback_1 = require("../../../base/feedback/css-feedback");
const tool_1 = require("../../../base/tool-manager/tool");
const viewpoint_util_1 = require("../../../utils/viewpoint-util");
const base_tools_1 = require("../base-tools");
const marquee_behavior_1 = require("./marquee-behavior");
const marquee_tool_feedback_1 = require("./marquee-tool-feedback");
const model_1 = require("../../../model");
let MarqueeMouseTool = MarqueeMouseTool_1 = class MarqueeMouseTool extends base_tools_1.BaseEditTool {
    constructor() {
        super(...arguments);
        this.shiftKeyListener = new ShiftKeyListener();
    }
    get id() {
        return MarqueeMouseTool_1.ID;
    }
    enable() {
        this.toDisposeOnDisable.push(this.mouseTool.registerListener(new MarqueeMouseListener(this.domHelper, this.editorContext.modelRoot, this.marqueeBehavior)), this.keyTool.registerListener(this.shiftKeyListener), this.registerFeedback([(0, css_feedback_1.cursorFeedbackAction)(css_feedback_1.CursorCSS.MARQUEE)], this, [(0, css_feedback_1.cursorFeedbackAction)()]));
    }
};
exports.MarqueeMouseTool = MarqueeMouseTool;
MarqueeMouseTool.ID = 'glsp.marquee-mouse-tool';
__decorate([
    (0, inversify_1.inject)(sprotty_1.TYPES.DOMHelper),
    __metadata("design:type", sprotty_1.DOMHelper)
], MarqueeMouseTool.prototype, "domHelper", void 0);
__decorate([
    (0, inversify_1.inject)(sprotty_1.TYPES.IMarqueeBehavior),
    (0, inversify_1.optional)(),
    __metadata("design:type", Object)
], MarqueeMouseTool.prototype, "marqueeBehavior", void 0);
exports.MarqueeMouseTool = MarqueeMouseTool = MarqueeMouseTool_1 = __decorate([
    (0, inversify_1.injectable)()
], MarqueeMouseTool);
class MarqueeMouseListener extends drag_aware_mouse_listener_1.DragAwareMouseListener {
    constructor(domHelper, root, marqueeBehavior) {
        super();
        this.isActive = false;
        this.domHelper = domHelper;
        this.marqueeUtil = new marquee_behavior_1.MarqueeUtil(marqueeBehavior);
        this.nodes = Array.from(root.index
            .all()
            .map(e => e)
            .filter(e => (0, sprotty_1.isSelectable)(e))
            .filter(e => e instanceof sprotty_1.GNode));
        const sEdges = Array.from(root.index
            .all()
            .filter(e => e instanceof model_1.GEdge)
            .filter(e => (0, sprotty_1.isSelectable)(e))
            .map(e => e.id));
        this.edges = Array.from(document.querySelectorAll('g')).filter(e => sEdges.includes(this.domHelper.findSModelIdByDOMElement(e)));
    }
    mouseDown(target, event) {
        this.isActive = true;
        this.marqueeUtil.updateStartPoint((0, viewpoint_util_1.getAbsolutePosition)(target, event));
        if (event.ctrlKey) {
            this.previouslySelected = Array.from(target.root.index
                .all()
                .map(e => e)
                .filter(e => (0, sprotty_1.isSelected)(e))
                .map(e => e.id));
        }
        return [];
    }
    mouseMove(target, event) {
        this.marqueeUtil.updateCurrentPoint((0, viewpoint_util_1.getAbsolutePosition)(target, event));
        if (this.isActive) {
            const nodeIdsSelected = this.nodes.filter(e => this.marqueeUtil.isNodeMarked((0, viewpoint_util_1.toAbsoluteBounds)(e))).map(e => e.id);
            const edgeIdsSelected = this.edges.filter(e => this.isEdgeMarked(e)).map(e => this.domHelper.findSModelIdByDOMElement(e));
            const selected = nodeIdsSelected.concat(edgeIdsSelected);
            return [sprotty_1.SelectAction.setSelection(selected.concat(this.previouslySelected)), this.marqueeUtil.drawMarqueeAction()];
        }
        return [];
    }
    mouseUp(_target, event) {
        this.isActive = false;
        if (event.shiftKey) {
            return [marquee_tool_feedback_1.RemoveMarqueeAction.create()];
        }
        return [marquee_tool_feedback_1.RemoveMarqueeAction.create(), tool_1.EnableDefaultToolsAction.create()];
    }
    isEdgeMarked(element) {
        if (!element.getAttribute('transform')) {
            if (element.children[0]) {
                const path = element.children[0].getAttribute('d');
                return this.marqueeUtil.isEdgePathMarked(path);
            }
        }
        return false;
    }
}
exports.MarqueeMouseListener = MarqueeMouseListener;
let ShiftKeyListener = class ShiftKeyListener extends sprotty_1.KeyListener {
    keyUp(element, event) {
        if (event.shiftKey) {
            return [];
        }
        return [marquee_tool_feedback_1.RemoveMarqueeAction.create(), tool_1.EnableDefaultToolsAction.create()];
    }
};
exports.ShiftKeyListener = ShiftKeyListener;
exports.ShiftKeyListener = ShiftKeyListener = __decorate([
    (0, inversify_1.injectable)()
], ShiftKeyListener);
