// @eclipse-glsp/sprotty/lib/feature-modules.js
"use strict";
/********************************************************************************
 * Copyright (c) 2023 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable no-restricted-imports */
Object.defineProperty(exports, "__esModule", { value: true });
exports.sprottyDefaultModule = exports.zorderModule = exports.updateModule = exports.openModule = exports.moveModule = exports.modelSourceModule = exports.labelEditUiModule = exports.fadeModule = exports.expandModule = exports.edgeLayoutModule = exports.edgeIntersectionModule = exports.edgeEditModule = exports.buttonModule = void 0;
const protocol_1 = require("@eclipse-glsp/protocol");
const di_config_1 = require("sprotty/lib/features/button/di.config");
const di_config_2 = require("sprotty/lib/features/edge-intersection/di.config");
const di_config_3 = require("sprotty/lib/features/edge-layout/di.config");
const di_config_4 = require("sprotty/lib/features/edit/di.config");
const di_config_5 = require("sprotty/lib/features/expand/di.config");
const di_config_6 = require("sprotty/lib/features/fade/di.config");
const di_config_7 = require("sprotty/lib/features/move/di.config");
const di_config_8 = require("sprotty/lib/features/update/di.config");
const di_config_9 = require("sprotty/lib/features/zorder/di.config");
const di_config_10 = require("sprotty/lib/model-source/di.config");
const di_config_11 = require("sprotty/lib/base/di.config");
exports.sprottyDefaultModule = di_config_11.default;
const di_config_12 = require("sprotty/lib/features/open/di.config");
exports.buttonModule = new protocol_1.FeatureModule(di_config_1.default.registry);
exports.edgeEditModule = new protocol_1.FeatureModule(di_config_4.edgeEditModule.registry);
exports.edgeIntersectionModule = new protocol_1.FeatureModule(di_config_2.default.registry);
exports.edgeLayoutModule = new protocol_1.FeatureModule(di_config_3.default.registry);
exports.expandModule = new protocol_1.FeatureModule(di_config_5.default.registry);
exports.fadeModule = new protocol_1.FeatureModule(di_config_6.default.registry);
exports.labelEditUiModule = new protocol_1.FeatureModule(di_config_4.labelEditUiModule.registry);
exports.modelSourceModule = new protocol_1.FeatureModule(di_config_10.default.registry);
exports.moveModule = new protocol_1.FeatureModule(di_config_7.default.registry);
exports.openModule = new protocol_1.FeatureModule(di_config_12.default.registry);
exports.updateModule = new protocol_1.FeatureModule(di_config_8.default.registry);
exports.zorderModule = new protocol_1.FeatureModule(di_config_9.default.registry);
